src/v/cluster/cluster_epoch_service.cc

```diff
@@ -0,0 +1,367 @@
+    auto cleanup = ss::defer([this]() noexcept { _current_epoch.reset(); });
+    while (!_abort_source.abort_requested()) {
+        try {
+            co_await ss::sleep_abortable(epoch_bump_interval, _abort_source);
+        } catch (const ss::sleep_aborted& e) {
+            std::ignore = e;
+        }
+        if (_abort_source.abort_requested()) {
+            co_return;
+        }
+        co_await update_epoch();
+    }
+}
+
+ss::future<> cluster_epoch_service::raft0_state::update_epoch() {
+    constexpr std::chrono::milliseconds base_backoff = 100ms;
+    constexpr std::chrono::milliseconds max_backoff = 10s;
+    constexpr std::chrono::milliseconds rpc_timeout = 3s;
+    auto policy = make_exponential_backoff_policy<ss::lowres_clock>(
+      base_backoff, max_backoff);
+    while (!_abort_source.abort_requested()) {
+        try {
+            co_await ss::sleep_abortable<ss::lowres_clock>(
+              policy.current_backoff_duration(), _abort_source);
+            policy.next_backoff(); // if we get a failure increment the backoff
+        } catch (const ss::sleep_aborted&) {
+            co_return;
+        }
+        auto deadline = ss::lowres_clock::now() + rpc_timeout;
+        auto result_fut
+          = co_await ss::coroutine::as_future<result<raft::replicate_result>>(
+            _stm.local().quorum_write_empty_batch(deadline));
+        if (result_fut.failed()) {
+            std::exception_ptr ex = result_fut.get_exception();
+            vlog(clusterlog.warn, "Failed to force increment epoch: {}", ex);
+            continue;
+        }
+        auto result = result_fut.get();
+        if (result) {
+            _current_epoch = result.value().last_offset;
+            vlog(
+              clusterlog.debug,
+              "Forced increment of cluster epoch to: {}",
+              _current_epoch);
+            co_return;
+        } else if (result.error() != raft::errc::not_leader) {
+            vlog(
+              clusterlog.warn,
+              "Failed to force increment epoch: {}",
+              result.error());
+        }
+    }
+}
+
+cluster_epoch_service::cluster_epoch_service(
+  model::node_id self,
+  ss::sharded<rpc::connection_cache>* cc,
+  ss::sharded<partition_leaders_table>* l) noexcept
+  : _self(self)
+  , _rpc_conn(cc)
+  , _leaders(l) {}
+
+cluster_epoch_service::~cluster_epoch_service() noexcept = default;
+
+ss::future<> cluster_epoch_service::start() { co_return; }
+ss::future<> cluster_epoch_service::stop() {
+    co_await _gate.close();
+    if (!_shard0_state) {
+        co_return;
+    }
+    co_await _shard0_state->stop();
+    _shard0_state = nullptr;
+}
+
+void cluster_epoch_service::invalidate_epoch_cache(int64_t observed_epoch) {
+    _gate.check();
+    // This is safe to check without the lock, because we only use the lock to
+    // limit the number of cross shard calls and RPCs.
+    if (_cached_epoch == observed_epoch) {
```

Comment:
shouldn't this be 

if (_cached_epoch < observed_epoch)

Reply:
I need to clarify which epoch this is. I was thinking you would pass in the local epoch that caused an out of sequence error, not whatever the higher sequence that was observed, because we really don't have a way to plumb this back on the write path (but maybe the STM knows how to invalidate the cache, which seems like a leak of abstraction possibly). I do think only passing around shard local epochs is better, but I need to clarify the docs a bunch.

Reply:
ah ok.. the name `observed_epoch` confused me.

Reply:
Renamed